File: zig-generator/src/emitters/mov_imm.zig
const std = @import("std");
const Emitter = @import("../emitter.zig").Emitter;
const GeneratorState = @import("../state.zig").GeneratorState;

pub const MovImmEmitter = struct {
    allocator: std.mem.Allocator,

    pub fn emitter(self: *MovImmEmitter) Emitter {
        return .{
            .ptr = self,
            .vtable = &.{ .try_emit = tryEmit, },
        };
    }

    fn tryEmit(ptr: *anyopaque, state: *GeneratorState) ?[]u8 {
        const self: *MovImmEmitter = @ptrCast(@alignCast(ptr));
        const dst = state.pickWritableGpr() orelse return null;
        const imm = state.getRng().uintAtMost(u32, 0xFFFF);
        state.markScalar(dst);

        return std.fmt.allocPrint(self.allocator, "  MOV x{d}, #0x{X:0>4}", .{dst, imm}) catch null;
    }
};

File: rust-generator/src/emitters/mov_imm.rs
use crate::emitter::Emitter;
use crate::state::GeneratorState;
use rand::RngExt;

pub struct MovImmEmitter;

impl Emitter for MovImmEmitter {
    fn try_emit(&mut self, state: &mut GeneratorState) -> Option<String> {
        let dst = state.pick_writable_gpr()?;
        let imm: u32 = state.rng().random_range(0..=0xFFFFu32);
        state.mark_scalar(dst);
        Some(format!("  MOV x{}, #0x{:04X}", dst, imm))
    }
}

